### ISSUE TITLE
[Docs] Installation needs refresh

### ISSUE THREAD

**Issue by jeromekelleher** [tskit-dev/msprime#994]
The installation page needs a refresh - Python versions are out of date, etc.

We should also mention how to run using the Docker image.

For HPC users, we might mention the [easybuild](https://easybuild.readthedocs.io/en/latest/version-specific/Supported_software.html?highlight=msprime#list-software-msprime-765) option. We'd probably need to learn how to update this recipe also (but I know one of the maintainers, so that shouldn't be too hard)

**Comment by benjeffery** [comment]
I'm assuming we intend to get wheels in for 1.0. At that point the build dependencies should only be needed in unusual cases.

**Comment by jeromekelleher** [comment]
Agreed - see #1004 for details tracking the progress of binary wheel building.

**Comment by castedo** [comment]
After reading this msprime installation page, there were a few changes I was thinking of making. And practicing my first PR on doc pages seems less dangerous than the C code. :-> Mainly I see some minor re-org that I think will make the instructions a little easy to understand and move the sub-instructions for super old platforms into one big "Old Platforms" section at the end.

While I'm at it, I can update the python version (I only see it in the conda output) and add docker instructions. I can also mention podman (which allows running containers as non-root, I use it).

I don't know easybuild so I won't attempt that right now.

It would make sense to move the "Containerzation" section out of the "Devloper documentation" page and into the "Installation" page right?

Is singularity still worth mentioning? This seems obsoleted by podman on RHEL (I can run msprime container without root with podman). Looks like podman will make it into Debian/Ubuntu next year.

Should I assign this bug to myself? (and I'll create a PR)

**Comment by jeromekelleher** [comment]
That sounds great, thanks @castedo. I'm not sure we need to keep all the stuff about installing on old platforms. With conda (and hopefully binary wheels, soon) there's less need to hand-hold people through linking up with GSL in different places. I'd say we just delete all that stuff.

The rest sounds great. I think there are quite a few people still using singularity (HPC shops move slowly) so I'd leave that in. I've just assigned the issue to you - have fun!

**Comment by castedo** [comment]
I'm getting ready to do a draft PR. I've done my first commit
https://github.com/castedo/msprime/commit/0b42c0e7e816113eaa55373d4eda6744161964e3
I have another commit to move container instructions from development.rst over into installation.rst.
And I plan to practice doing a rebase/squash for the first time before un-drafting the PR.

While doing all this, I couldn't help but ponder some high level questions.
Perhaps answering these questions and recording them into the wiki could be
useful in planning and making trade-offs and explaining decisions.

The question that came to my mind is:

How are msprime users split in terms of installation needs?

1) "Desktop User":
- using msprime in research
- running on Windows, Mac or Linux
- "research code" is _NOT_ tested/used on platforms other than the user's desktop
- ? likely running research code in a Jupyter notebook
- ? circa 2020, not knowledgable in running VMs or containers to run Linux software on Windows or Mac (but maybe in 2021/2020)

2) "Software Developer"
- software written by "developer" _IS_ tested/used on multiple platforms
- ? developers on Windows are not a target profile to support?
- ? can Mac developers be expected to developing within Linux containers?
- "developer" further trisected into
  a) "msprime developers" (changing msprime)
  b) "Python-only developers"
       - building on top of msprime and other Python packages
       - ? likely pip is preferred Python package manager ?
  c) "Multi-language developers"
      - building on top of msprime and other non-Python dependencies
      - ? likely system package managers are preferred (i.e. yum, apt-get)?

3) "Server Admin":
- running msprime on servers, say HPC
- ? happy for msprime to run within a container?

Let me know if you think consensus answers to these questions in this break down would be useful to be the wiki (or other planning area).

And I'm kind of curious to some answers regardless of whether or not they get recorded elsewhere.

**Comment by agladstein** [comment]
So... funny thing. I'm trying to verify https://github.com/tskit-dev/msprime/pull/1136, but I can't get the dev version installed locally so that I can build @castedo new docs. (I'm on a Mac)

Following the current instillation docs is not doing me any good.
This was my attempt (in a virtualenv):
```
python3 -m venv msprimedev_env
source msprimedev_env/bin/activate
python3 -m pip  install --upgrade pip setuptools
python3 -m pip install -r requirements/development.txt
```

When I try to make the docs, I get this:
```
(msprimedev_env) Ariellas-New-MacBook-Pro:docs agladsteinNew$ make
python3 -m sphinx -b html -d _build/doctrees  -W . _build/html
Running Sphinx v3.2.1
loading intersphinx inventory from https://docs.python.org/3/objects.inv...
loading intersphinx inventory from http://docs.scipy.org/doc/numpy/objects.inv...
loading intersphinx inventory from https://tskit.readthedocs.io/en/stable/objects.inv...
loading intersphinx inventory from https://stdpopsim.readthedocs.io/en/stable/objects.inv...
intersphinx inventory has moved: http://docs.scipy.org/doc/numpy/objects.inv -> https://numpy.org/doc/stable/objects.inv
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 9 source files that are out of date
updating environment: [new config] 9 added, 0 changed, 0 removed
reading sources... [100%] tutorial

Warning, treated as error:
autodoc: failed to import function 'simulate' from module 'msprime'; the following exception was raised:
cannot import name '_msprime' from 'msprime' (/Users/agladsteinNew/dev/msprime/msprime/__init__.py)
make: *** [html] Error 2
```

**Comment by benjeffery** [comment]
I think you need to run `make` at the root first to build the extension.

**Comment by agladstein** [comment]
> How are msprime users split in terms of installation needs?

I agree with your breakup. I think "Server Admin" also extends to cloud. People can run distributed simulations on the cloud (or something like the Open Science Grid).

> Is singularity still worth mentioning? This seems obsoleted by podman on RHEL (I can run msprime container without root with podman). Looks like podman will make it into Debian/Ubuntu next year.

I've never heard of podman! But, I know that a lot of HPC systems use singularity and therefore, having singularity instructions is great for HPC.

**Comment by agladstein** [comment]
> I think you need to run make at the root (one level up from docs) first to build the extension.

I'm getting this:
```
(msprimedev_env) Ariellas-New-MacBook-Pro:msprime agladsteinNew$ make
python3 setup.py build_ext --inplace
your setuptools is too old (<12)
setuptools_scm functionality is degraded
running build_ext
building 'msprime._msprime' extension
gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -g -arch x86_64 -UNDEBUG -Ilib -Ilib/subprojects/tskit -Ilib/subprojects/tskit/tskit -Ilib/subprojects/kastore -I/usr/local/Cellar/gsl/2.6/include -I/Users/agladsteinNew/dev/msprime/msprimedev_env/include -I/Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m -I/Users/agladsteinNew/dev/msprime/msprimedev_env/lib/python3.7/site-packages/numpy/core/include -c msprime/_msprimemodule.c -o build/temp.macosx-10.9-x86_64-3.7/msprime/_msprimemodule.o -std=c99
In file included from msprime/_msprimemodule.c:33:
lib/msprime.h:27:10: fatal error: 'tskit.h' file not found
#include <tskit.h>
         ^~~~~~~~~
1 error generated.
error: command 'gcc' failed with exit status 1
make: *** [ext3] Error 1
```

I ran `brew install gsl` and same error.
I also did `pip3 install --upgrade pip setuptools`

**Comment by agladstein** [comment]
The installation refresh should also include this https://github.com/tskit-dev/msprime/issues/817

**Comment by benjeffery** [comment]
`git submodule update --init` maybe? Then `make`?

**Comment by castedo** [comment]
@agladstein that first error you got with not being able to import _msprime I started to get after the recent merge which moved _msprime into msprime. I was able to work around the issue locally by replacing "_msprime" in conf.py with "msprime._msprime".

**Comment by castedo** [comment]
I haven't had a chance to isolate exactly why I could build the docs before this week's merge but not afterwards (except with the work around I just mentioned). I suspect that the docs are using some clever object mocking trick so that not everything in msprime needs to be build (especially the C stuff). If my hunch is right, it would be a nice trick because then "incomplete" local setups could still build the docs even though other stuff is not working outside docs. I suspect the reason the CI doc test is passing is because the mock trick for docs is avoided if a full msprime setup is installed and working.

**Comment by castedo** [comment]
the PR/merge I'm referring to is #1133. When I roll back to right before that merge I was able to build the docs with an ancient v2 Sphinx and a bunch of weird RedHat non-pip older versions of various Python packages. Then when I fast-foward to the marge is breaks and I needed to do the above work around.
So I'm thinking I should just add this work around to line 34 of the conf.py even though I haven't isolated the exact cause.

**Comment by jeromekelleher** [comment]
> @agladstein that first error you got with not being able to import _msprime I started to get after the recent merge which moved _msprime into msprime. I was able to work around the issue locally by replacing "_msprime" in conf.py with "msprime._msprime".

Well spotted, thanks @castedo - I've made a PR to fix this in #1140.

**Comment by castedo** [comment]
My first PR has been merged! Yay!
So what's the usual process for an github issue associated with a PR, like this one. It's assigned to me. Who decides whether an issue is closed? I see everything in this issue has been either addressed or moved into #1144.
